### ISSUE TITLE
How to access error responses

### ISSUE THREAD

**Issue by mbeynon** [balanced/balanced-node#34]
I'd like to process the error response of a balanced debit call without resorting to string processing.  The below code is referencing the sample card number on the balanced website that simulates a processor failure (last four is 4448), so it should fail.

``` javascript
balanced.get(myCardHref).debit({ amount: 1000})
      .then(function(obj) {
        // handle success
      })
      .catch(function(err) {
        console.log(err);
      });
```

The err object in the reject clause appears to be an instance of Error.  How can I instead get access to the actual response object so I can access the various "errors" and "debits" clauses in the jsonapi result?  I assume there must be something cleaner than JSON.parse(err["message"]).

``` javascript
[Error: {
  "errors": [
    {
      "status": "Payment Required",
      "category_code": "card-declined",
      "additional": "Account Frozen",
      "status_code": 402,
      "category_type": "banking",
      "extras": {},
      "request_id": "OHM431eec14e22111e3be3f02b12035401b",
      "description": "R758: Account Frozen. Your request id is OHM431eec14e22111e3be3f02b12035401b."
    }
  ],
  "links": {
    "debits.customer": "/customers/{debits.customer}",
    "debits.dispute": "/disputes/{debits.dispute}",
    "debits.source": "/resources/{debits.source}",
    "debits.order": "/orders/{debits.order}",
    "debits.refunds": "/debits/{debits.id}/refunds",
    "debits.events": "/debits/{debits.id}/events"
  },
  "debits": [
    {
      "status": "failed",
      "transaction_number": "W877-022-7997",
      "description": "SeeYourGuru.com charge",
      "links": {
        "customer": "CU3amMvcf0DcY1sEVJuscRxj",
        "source": "CC3hqWNlPglj3hYp0Uw8ktF3",
        "order": null,
        "dispute": null
      },
      "created_at": "2014-05-23T02:23:43.977581Z",
      "updated_at": "2014-05-23T02:23:44.261873Z",
      "failure_reason": "R758: Account Frozen.",
      "currency": "USD",
      "amount": 4300,
      "href": "/debits/WD22IHTvI6QDD64Z1VL9IfJj",
      "meta": {},
      "failure_reason_code": "card-declined",
      "appears_on_statement_as": "BAL*Guru charge",
      "id": "WD22IHTvI6QDD64Z1VL9IfJj"
    }
  ]
}]
```

**Comment by matthewfl** [comment]
https://github.com/matthewfl/jsonapi-node/blob/master/lib.js#L89

There is no good way to do this currently, but as you said, there should be a good way to get at the error response without doing any hacks

**Comment by mbeynon** [comment]
I'd like to fix.  Which of the following would you prefer?

(1) modify jsonapi.prototype._req to return a derived JsonReqError object that adds a json_response property to hold the JSON object return.

(2) modify jsonapi.prototype._req statements from "ret.reject(e);" to pass in a second argument to become "ret.reject(e,jsonResponse);".

**Comment by mjallday** [comment]
@matthewfl is there an accepted style for writing the error handlers? it seems more natural for the response to be an attribute on the exception if it's derived from a base http exception.

**Comment by matthewfl** [comment]
@mbeynon I think that approach 1 would be best for this case.
@mjallday that would be best, but there is currently no hierarchy of errors, I am also unsure if there are any good examples of this being used in javascript since they try-cache will end up catching any exception

**Comment by matthewfl** [comment]
https://github.com/balanced/balanced-node/blob/master/test/test.js#L467-L480

The errors that are returned now have the fields from the error message returned by balanced.  This means you can do `err.description` and get the human readable error message returned by balanced.

Fixed in v1.3.0
